pkg/optimize/acl/reduceCubes.go

```diff
@@ -8,53 +8,178 @@ package acloptimizer
+		r := ds.CartesianPairLeft(p.S1, p.S2)
+		anyICMP = anyICMP.Union(r).(*srcDstProductLeft)
 		if p.S3.IsAll() { // all icmp types and codes
-			r := ds.CartesianPairLeft(p.S1, p.S2)
-			res = res.Union(r).(*srcDstProductLeft)
+			allICMP = allICMP.Union(r).(*srcDstProductLeft)
 		}
 	}
-	return res
+	return
 }
 
+// Updates the cubes so that if for <src, dst> the other two protocols are complete (for tcp/udp all
+// ports and for icmp all types&codes) and the complement of the third protocol can be written by one rule
+// we convert it to two rules, deny complement and allow any
+func excludeICMP(cubes *aclCubesPerProtocol, allTCPUDP srcDstProduct) {
+	var icmpSet *netset.ICMPSet
+	var single bool
+	for _, p := range cubes.icmpAllow.Partitions() {
+		if icmpSet, single = singleICMPComplementValue(p.S3); !single {
+			continue
+		}
+		intersectedCube := ds.CartesianPairLeft(p.S1, p.S2).Intersect(allTCPUDP)
+		cubes.icmpAllow = subtractSrcDstCubeFromICMP(cubes.icmpAllow, intersectedCube, p.S3)
+		cubes.icmpDeny = addSrcDstCubeToICMP(cubes.icmpDeny, intersectedCube, icmpSet)
+		cubes.anyProtocolAllow = cubes.anyProtocolAllow.Union(intersectedCube).(*srcDstProductLeft)
+	}
+}
+
+// returns whether the complement of the given set can be written in one rule (and the complement set)
+func singleICMPComplementValue(icmpSet *netset.ICMPSet) (*netset.ICMPSet, bool) {
+	complementSet := netset.AllICMPSet().Subtract(icmpSet)
+	return complementSet, len(optimize.IcmpsetPartitions(complementSet)) == 1
+}
+
+// subtractAnyProtocolCubes from each protocol
 func subtractAnyProtocolCubes(aclCubes *aclCubesPerProtocol) {
-	allTcpudp := ds.NewLeftTripleSet[*netset.IPBlock, *netset.IPBlock, *netset.TCPUDPSet]()
-	allIcmp := ds.NewLeftTripleSet[*netset.IPBlock, *netset.IPBlock, *netset.ICMPSet]()
-	for _, p := range aclCubes.anyProtocolAllow.Partitions() {
-		t := ds.CartesianLeftTriple(p.Left, p.Right, netset.AllTCPUDPSet())
-		allTcpudp = allTcpudp.Union(t).(*ds.LeftTripleSet[*netset.IPBlock, *netset.IPBlock, *netset.TCPUDPSet])
-		i := ds.CartesianLeftTriple(p.Left, p.Right, netset.AllICMPSet())
-		allIcmp = allIcmp.Union(i).(*ds.LeftTripleSet[*netset.IPBlock, *netset.IPBlock, *netset.ICMPSet])
-	}
-
-	aclCubes.tcpAllow = aclCubes.tcpAllow.Subtract(allTcpudp)
-	aclCubes.udpAllow = aclCubes.udpAllow.Subtract(allTcpudp)
-	aclCubes.icmpAllow = aclCubes.icmpAllow.Subtract(allIcmp)
+	aclCubes.tcpAllow = subtractSrcDstCubeFromTCPUDP(aclCubes.tcpAllow, aclCubes.anyProtocolAllow, netset.NewAllTCPOnlySet())
+	aclCubes.udpAllow = subtractSrcDstCubeFromTCPUDP(aclCubes.udpAllow, aclCubes.anyProtocolAllow, netset.NewAllUDPOnlySet())
+	aclCubes.icmpAllow = subtractSrcDstCubeFromICMP(aclCubes.icmpAllow, aclCubes.anyProtocolAllow, netset.AllICMPSet())
+}
+
+// Union <src X dst> cube with pr protocolSet to tcpudp cube
+func addSrcDstCubeToTCPUDP(tcpudpCube tcpudpTripleSet, srcDstCube srcDstProduct, pr *netset.TCPUDPSet) tcpudpTripleSet {
```

Comment:
Can we use generics/interfaces to have just one `add` and one `subtract` for both TCPUDP and for ICMP?